### PR TITLE
AMMOSGH-115: Added implementation for Kerberos and Keytab File based …

### PR DESCRIPTION
**Pull request by IbraheemYSaleh**
…login methods for CAM support

### CODE REVIEW

**Comment by IbraheemYSaleh** [comment]
Not ready for merge yet -- Needs further testing & verification but our integration testbed that has CAM security setup is down for kickstarts for the next day or so. It should be ready for review.

File: src/src_cryptography/src_kmc_crypto_service/cryptography_interface_kmc_crypto_service.template.c [hunk 1]
```diff
@@ -1925,6 +1964,85 @@ int32_t curl_response_error_check(CURL* curl_handle, char* response)
...
+    if(cam_config->username == NULL)
+    {
+        status = CAM_INVALID_CONFIGURATION_KEYTAB_FILE_USERNAME_NULL;
+        return status;
+    }
+
+    // Build the kinit shell command with keytab file path + username
+    char* kinit_shell_command_base = "kinit -kt %s %s";
+    uint32_t len_kinit_shell_command = strlen(kinit_shell_command_base) + strlen(cam_config->keytab_file_path) + strlen(cam_config->username);
+    char* kinit_shell_command = malloc(len_kinit_shell_command) + 1;
```
Comment (rjbrown2) [inline src/src_cryptography/src_kmc_crypto_service/cryptography_interface_kmc_crypto_service.template.c:2034]: Are these freed somewhere that I'm missing?

File: src/src_cryptography/src_kmc_crypto_service/cryptography_interface_kmc_crypto_service.template.c [hunk 2]
```diff
@@ -1860,6 +1793,103 @@ static int32_t handle_cam_cookies(CURL* curl_handle, char* cam_cookies)
...
+    {
+        curl_easy_setopt(curl_cam, CURLOPT_USERNAME, cam_config->username);
+    } else
+    {
+        curl_easy_setopt(curl_cam, CURLOPT_USERNAME, ":");
+    }
+
+    // Build the CAM getSsoToken URI
+    int len_kerberos_endpoint = strlen(cam_kerberos_uri) + strlen(cam_config->access_manager_uri);
+    char* kerberos_endpoint_final = (char*) malloc(len_kerberos_endpoint) + 1;
```
Comment (rjbrown2) [inline src/src_cryptography/src_kmc_crypto_service/cryptography_interface_kmc_crypto_service.template.c:1858]: Free me?

**Review by rjbrown2 — COMMENTED** [review]
Other than the couple of mallocs, I think this looks ok.  Let us know when you've finished and are ready to merge.

**Comment by IbraheemYSaleh** [comment]
Thanks for reviewing @rjbrown2 ! I'm still waiting on the CAM secured testbed to be ready to properly test this... I'll address the mallocs at the same time I test & verify this all works properly... Thanks again!

File: src/src_cryptography/src_kmc_crypto_service/cryptography_interface_kmc_crypto_service.template.c (hunk 1 above)
Comment (IbraheemYSaleh) [reply src/src_cryptography/src_kmc_crypto_service/cryptography_interface_kmc_crypto_service.template.c:2082]: Good catch! I've added the corollary free for this malloc.

**Comment by IbraheemYSaleh** [comment]
Our CAM-enabled testbed is now setup and I've now fixed up and tested this code. It should be ready for merge. 

CC @rjbrown2

**Review by rjbrown2 — APPROVED** [review]
Looks OK to me!

**Pull request merged**